manager/src/main/java/org/openremote/manager/map/MapService.java

```diff
@@ -377,6 +489,116 @@ public byte[] getMapTile(int zoom, int column, int row) {
+
+        // Check there's no back refs in the filename
+        boolean isValid = tilesPath.toFile().getCanonicalPath().contains(customTilesDir.toFile().getCanonicalPath() + File.separator);
+        if (!isValid) {
+            String msg = "Filename outside permitted directory: " + filename;
+            LOG.warning(msg);
+            throw new IllegalArgumentException(msg);
+        }
+
+        try (OutputStream outputStream = Files.newOutputStream(tilesPath)) {
+            byte[] buffer = new byte[4096];
+            int bytesRead;
+            int written = 0;
+            while ((bytesRead = fileInputStream.read(buffer)) != -1) {
+                if (written > customMapLimit) {
+                    String msg = "Stream continued passed custom map limit size: " + tilesPath;
+                    LOG.log(Level.SEVERE, msg);
+                    throw new IOException(msg);
+                }
+                outputStream.write(buffer, 0, bytesRead);
+                written += bytesRead;
+            }
+        } catch (IOException e) {
+            try {
+                // Attempt to delete this invalid map data file
+                Files.deleteIfExists(tilesPath);
+            } catch (IOException ex) {
+                LOG.log(Level.WARNING, "Could not delete partially written file: " + filename, ex);
+            }
+            throw e;
+        }
+
+        // Ensure we can access this new file and that it is a valid mbtiles file
+        Path loadedFile = setData(false);
+        // TODO: Is this save necessary?
+        saveMapMetadata(metadata);
+
+        if (loadedFile == null || !loadedFile.toAbsolutePath().equals(tilesPath.toAbsolutePath())) {
+            try {
+                // Attempt to delete this invalid map data file
+                Files.deleteIfExists(tilesPath);
+            } catch (IOException ex) {
+                LOG.log(Level.WARNING, "Could not delete partially written file: " + filename, ex);
+            }
+            throw new IOException("Failed to load map data ensure the uploaded file is a valid mbtiles file: " + filename);
+        }
+
+        // Now delete any previous custom mbtiles file
+        if (previousCustomTilesPath.toFile().isFile()) {
+            try {
+                // Attempt to delete this old map data file
+                Files.deleteIfExists(previousCustomTilesPath);
+            } catch (IOException ex) {
+                LOG.log(Level.WARNING, "Could not delete old file: " + previousCustomTilesPath, ex);
+                // This is problematic as which custom file gets discovered first on next startup??
+            }
+        }
+    }
+
+    public void deleteUploadedFile() throws IOException {
+        Path previousCustomTilesPath = configurationService.getCustomMapTilesPath(true);
+
+        if (previousCustomTilesPath.toFile().isFile()) {
+
+            setData(true);
+            // TODO: Is this save necessary?
+            saveMapMetadata(metadata);
```

Comment:
Address this TODO statement; is this necessary?
Put a simple context message in the code I'd say.